### PR TITLE
Push spraggins to google cloud

### PR DESCRIPTION
**Pull request by manzt**
The linnarsson update in `v.0.23` ignores all `.zarr` files from syncing with AWS. This is my mistake. We haven't been using the Spraggins dataset in Vitessce (yet), but this change prevented `spraggins.ims.zarr` from being pushed in the most recent release. 

This PR updates the destination URL for the spraggins IMS dataset to google cloud storage, explicitly pushes the zarr array there, and updates some oddities that we've run into trying to get these data to render in Viv. 

- Adds a `transform` and `domain` keys to `spraggins.ims.json`. 
-- `transform`: includes `scale` and `translate` keys to position image in renderer. This could include other transformations in the future. 
-- `domain` describes the domain of values in the datasets, generally `[0, max]`
- Changed the `dtype` to `uint32` rather than `int32`

I think we should move towards parsing the zarr metadata in the future, so I have mirrored the changes in `spraggins.ims.json` to `.zattrs` and we can work towards this in the future.

### CODE REVIEW

File: fake-files/output-expected/spraggins/spraggins.ims.json [hunk 1]
```diff
@@ -5,7 +5,18 @@
     "x"
   ],
   "zarrConfig": {
-    "store": "https://s3.amazonaws.com/vitessce-data/0.0.24/master_release/spraggins/",
+    "store": "https://vitessce-data.storage.googleapis.com/0.0.24/master_release/spraggins",
     "path": "spraggins.ims.zarr"
+  },
+  "domain": [
```
Comment (ilan-gold) [inline fake-files/output-expected/spraggins/spraggins.ims.json:11]: Can we also change https://github.com/hubmapconsortium/vitessce-data/blob/master/python/img_hdf5_reader.py#L195 while we do this one?

File: python/imzml_reader.py [hunk 2]
```diff
@@ -183,10 +193,22 @@ def write_metadata_json(json_file):
...
+        help="Destination for zarr output in cloud.",
+    )
     args = parser.parse_args()
 
     dataset = IMSDataset(
         args.imzml_file, args.ibd_file, micro_res=0.5, ims_res=10
     )
-    dataset.write_zarr(args.ims_zarr, dtype="i4", compressor=Zlib(level=1))
-    write_metadata_json(args.ims_metadata)
+    dataset.write_zarr(args.ims_zarr, dtype="uint32", compressor=Zlib(level=1))
```
Comment (ilan-gold) [inline python/imzml_reader.py]: Is there any way to infer `dtype` from the data?  If so, I think it would be good so that this could be a bit more robust.

**Review by ilan-gold — APPROVED** [review]

**Comment by ilan-gold** [comment]
I like the mirroring of the changes, thanks for that.

File: process.sh [hunk 3]
```diff
@@ -37,6 +37,7 @@ main() {
     else
         aws s3 cp --exclude "$OUTPUT/*.ome.tif*" --exclude "$OUTPUT/*.zarr/*" --recursive "$OUTPUT" s3://"$CLOUD_TARGET"
         gsutil -m cp -r "$OUTPUT/vanderbilt/vanderbilt.images" "gs://$CLOUD_TARGET/vanderbilt/vanderbilt.images"
+        gsutil -m cp -r "$OUTPUT/spraggins/spraggins.ims.zarr" "gs://$CLOUD_TARGET/spraggins/spraggins.ims.zarr"
         gsutil -m cp -r "$OUTPUT/linnarsson/linnarsson.images.zarr" "gs://$CLOUD_TARGET/linnarsson/linnarsson.images.zarr"
```
Comment (mccalluc) [inline process.sh:41]: This is starting to get cluttered. Two possible directions:
- Put this in a for-loop, like
```suggestion
        for DIR in vanderbilt/vanderbilt.images spraggins/spraggins.ims.zarr linnarsson/linnarsson.images.zarr; do
          gsutil -m cp -r "$OUTPUT/$DIR" "gs://$CLOUD_TARGET/$DIR"
        done
```
- or, perhaps better, look into [`gsutil rsync`](https://cloud.google.com/storage/docs/gsutil/commands/rsync), which does have an exclude option.

File: python/imzml_reader.py [hunk 4]
```diff
@@ -139,20 +153,28 @@ def write_zarr(self, path, dtype, compressor=None, chunks=None):
...
         z_arr[:, :, :] = arr
-        z_arr.attrs["x_extent"] = [float(extent.x_min), float(extent.x_max)]
-        z_arr.attrs["y_extent"] = [float(extent.y_min), float(extent.y_max)]
-        z_arr.attrs["scaling_factor"] = self.ims_px_in_micro
+        self.domain = [int(np.min(z_arr)), int(np.max(z_arr))]
+        self.translate = [
+            int(extent.y_min * self.ims_px_in_micro),
+            int(extent.x_min * self.ims_px_in_micro),
+        ]
+        z_arr.attrs["domain"] = self.domain
```
Comment (mccalluc) [inline python/imzml_reader.py:161]: Do we use self.domain anywhere else? Consider assigning directly.

Comment (mccalluc) [reply python/imzml_reader.py:161]: Nevermind, it's in write_metadata_json, I think.

File: python/imzml_reader.py [hunk 5]
```diff
@@ -139,20 +153,28 @@ def write_zarr(self, path, dtype, compressor=None, chunks=None):
...
+):
     json_out = {
         "dimensions": ["mz", "y", "x"],
-        "zarrConfig": {
-            "store": f"https://s3.amazonaws.com/{cloud_target}/spraggins/",
-            "path": "spraggins.ims.zarr",
-        },
+        "zarrConfig": {"store": zarr_store_url, "path": zarr_path},
+        "domain": domain,
+        "transform": {"scale": scale, "translate": translate},
```
Comment (mccalluc) [inline python/imzml_reader.py:177]: It looks like you do a similar thing for z_arr.attrs... perhaps the transform could be passed around, and not just the components?

File: scripts/process_spraggins.sh [hunk 6]
```diff
@@ -13,7 +13,8 @@ main() {
     ZARR_OUT="$OUTPUT/spraggins.ims.zarr"
     JSON_OUT="$OUTPUT/spraggins.ims.json"
 
-    CLI_ARGS="--imzml_file $IMZML_IN --ibd_file $IBD_IN --ims_zarr $ZARR_OUT --ims_metadata $JSON_OUT"
+    RELEASE=${CLOUD_TARGET//vitessce-data\//}
```
Comment (mccalluc) [inline scripts/process_spraggins.sh:16]: Why the backslash?

**Review by mccalluc — APPROVED** [review]
A couple questions. We should find a way to keep process.sh tidy, but it doesn't need to be this PR.

File: scripts/process_spraggins.sh (hunk 6 above)
Comment (manzt) [reply scripts/process_spraggins.sh:16]:
```
echo ${CLOUD_TARGET//vitessce-data//}
//0.0.24/master_release
```

```
echo ${CLOUD_TARGET//vitessce-data\//}
0.0.24/master_release
```

Comment (mccalluc) [reply scripts/process_spraggins.sh:16]: Cool: I wasn't remembering the replacement syntax. Sorry!

**Pull request merged**